### PR TITLE
Adding Attention and MatrixAttention modules

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

File: allennlp/modules/matrix_attention.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    Parameters
+    ----------
+    similarity_function: ``SimilarityFunction``, optional (default=``DotProductSimilarity``)
+        The similarity function to use when computing the attention.
+    '''
+    def __init__(self, similarity_function: SimilarityFunction = None) -> None:
+        super(MatrixAttention, self).__init__()
+
+        # This import is here to avoid circular import problems with the registry...
+        from allennlp.modules.similarity_functions.dot_product import DotProductSimilarity
```
Comment (DeNeutoy) [inline allennlp/modules/matrix_attention.py]: We need to have a better solution than this. This is unsustainable and confusing for contributors in the future - what that solution is, I am not exactly sure though ...

File: tests/common/tensor_test.py [hunk 2]
```diff
@@ -94,3 +97,110 @@ def test_sort_tensor_by_length(self):
 
         # Test restoration indices correctly recover the original tensor.
         assert sorted_tensor[reverse_indices].equal(tensor)
+
+    def test_masked_softmax_no_mask(self):
```
Comment (DeNeutoy) [inline tests/common/tensor_test.py]: Great tests 👍

File: allennlp/modules/matrix_attention.py (hunk 1 above)
Comment (matt-gardner) [inline allennlp/modules/matrix_attention.py]: Yeah, this is the main problem with how the `Registry` is set up.  I'm not sure how to fix this in a reasonable way.  @joelgrus, any ideas?

File: tests/common/tensor_test.py (hunk 2 above)
Comment (matt-gardner) [inline tests/common/tensor_test.py]: It's actually probably a bit excessive...  =).  I just copied it from DeepQA - Nelson wrote them original, and I just translated it to pytorch.

**Pull request merged**